sklearn/impute.py

```diff
@@ -873,3 +880,309 @@ def fit(self, X, y=None):
+        neighbors search because they have too many ( > row_max_missing)
+        missing features or because all of the sample's k-nearest neighbors
+        (i.e., the potential donors) also have the relevant feature value
+        missing.
+
+    References
+    ----------
+    * Olga Troyanskaya, Michael Cantor, Gavin Sherlock, Pat Brown, Trevor
+      Hastie, Robert Tibshirani, David Botstein and Russ B. Altman, Missing
+      value estimation methods for DNA microarrays, BIOINFORMATICS Vol. 17
+      no. 6, 2001 Pages 520-525.
+
+    Examples
+    --------
+    >>> from sklearn.impute import KNNImputer
+    >>> nan = float("NaN")
+    >>> X = [[1, 2, nan], [3, 4, 3], [nan, 6, 5], [8, 8, 7]]
+    >>> imputer = KNNImputer(n_neighbors=2, weights="uniform")
+    >>> imputer.fit_transform(X)
+    array([[1. , 2. , 4. ],
+           [3. , 4. , 3. ],
+           [5.5, 6. , 5. ],
+           [8. , 8. , 7. ]])
+    """
+
+    def __init__(self, missing_values="NaN", n_neighbors=5,
+                 weights="uniform", metric="masked_euclidean",
+                 row_max_missing=0.5, col_max_missing=0.8, copy=True):
+
+        self.missing_values = missing_values
+        self.n_neighbors = n_neighbors
+        self.weights = weights
+        self.metric = metric
+        self.row_max_missing = row_max_missing
+        self.col_max_missing = col_max_missing
+        self.copy = copy
+
+    def _impute(self, dist, X, fitted_X, mask, mask_fx):
+        """Helper function to find and impute missing values"""
+
+        # For each column, find and impute
+        n_rows_X, n_cols_X = X.shape
+        for c in range(n_cols_X):
+            if not np.any(mask[:, c], axis=0):
+                continue
+
+            # Row index for receivers and potential donors (pdonors)
+            receivers_row_idx = np.where(mask[:, c])[0]
+            pdonors_row_idx = np.where(~mask_fx[:, c])[0]
+
+            # Impute using column mean if n_neighbors are not available
+            if len(pdonors_row_idx) < self.n_neighbors:
+                warnings.warn("Insufficient number of neighbors! "
+                              "Filling in column mean.")
+                X[receivers_row_idx, c] = self.statistics_[c]
+                continue
+
+            # Get distance from potential donors
+            dist_pdonors = dist[receivers_row_idx][:, pdonors_row_idx]
+            dist_pdonors = dist_pdonors.reshape(-1,
+                                                len(pdonors_row_idx))
+
+            # Argpartition to separate actual donors from the rest
+            pdonors_idx = np.argpartition(
+                dist_pdonors, self.n_neighbors - 1, axis=1)
+
+            # Get final donors row index from pdonors
+            donors_idx = pdonors_idx[:, :self.n_neighbors]
+
+            # Get weights or None
+            dist_pdonors_rows = np.arange(len(donors_idx))[:, None]
+            weight_matrix = _get_weights(
+                dist_pdonors[
```

Comment:
we tend to avoid this indent-after-bracket style, preferring the readability of:

```
            weight_matrix = _get_weights(dist_pdonors[dist_pdonors_rows,
                                                      donors_idx],
                                         self.weights)
```